ipnipublisher/store/store.go

```diff
@@ -293,40 +298,45 @@ func Entries(ctx context.Context, ds Store, root ipld.Link) iter.Seq2[multihash.
 }
 
 func Encode(ctx context.Context, ds Store, lnk ipld.Link, w io.Writer) error {
-	r, err := ds.Get(ctx, lnk.String())
+	v, err := ds.Get(ctx, lnk.String())
 	if err != nil {
 		return err
 	}
-	defer r.Close()
-	_, err = io.Copy(w, r)
+	_, err = w.Write(v)
 	return err
 }
 
 func Head(ctx context.Context, ds Store) (*head.SignedHead, error) {
-	r, err := ds.Get(ctx, headKey)
+	v, err := ds.Get(ctx, headKey)
 	if err != nil {
 		return nil, err
 	}
-	defer r.Close()
-	return head.Decode(r)
+	return head.Decode(bytes.NewReader(v))
 }
 
 func EncodeHead(ctx context.Context, ds Store, w io.Writer) error {
-	r, err := ds.Get(ctx, headKey)
+	v, err := ds.Get(ctx, headKey)
 	if err != nil {
 		return err
 	}
-	defer r.Close()
-	_, err = io.Copy(w, r)
+	_, err = w.Write(v)
 	return err
 }
 
-func PutHead(ctx context.Context, ds Store, newHead *head.SignedHead) (datamodel.Link, error) {
+func ReplaceHead(ctx context.Context, ds Store, oldHead *head.SignedHead, newHead *head.SignedHead) (datamodel.Link, error) {
 	blk, err := block.Encode(newHead, head.SignedHeadPrototype.Type(), json.Codec, sha256.Hasher)
 	if err != nil {
 		return nil, err
 	}
-	err = ds.Put(ctx, headKey, uint64(len(blk.Bytes())), bytes.NewReader(blk.Bytes()))
+	var oldBytes []byte
+	if oldHead != nil {
+		blk, err := block.Encode(oldHead, head.SignedHeadPrototype.Type(), json.Codec, sha256.Hasher)
+		if err != nil {
+			return nil, err
+		}
+		oldBytes = blk.Bytes()
+	}
+	err = ds.Replace(ctx, headKey, oldBytes, blk.Bytes())
```

Comment:
I'd do `newBytes := blk.Bytes()` at the top after encoding the new head and then here `err = ds.Replace(ctx, headKey, oldBytes, newBytes)`. I was confused for a moment because `blk` is shadowed in the scope where the old head is encoded.